include/secp256k1_schnorrsig.h

```diff
@@ -0,0 +1,110 @@
+#ifndef SECP256K1_SCHNORRSIG_H
+#define SECP256K1_SCHNORRSIG_H
+
+#include "secp256k1.h"
+#include "secp256k1_extrakeys.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** This module implements a variant of Schnorr signatures compliant with
+ *  Bitcoin Improvement Proposal 340 "Schnorr Signatures for secp256k1"
+ *  (https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki).
+ */
+
+/** A pointer to a function to deterministically generate a nonce.
+ *
+ *  Same as secp256k1_nonce function with the exception of accepting an
+ *  additional pubkey argument and not requiring an attempt argument. The pubkey
+ *  argument can protect signature schemes with key-prefixed challenge hash
+ *  inputs against reusing the nonce when signing with the wrong precomputed
+ *  pubkey.
+ *
+ *  Returns: 1 if a nonce was successfully generated. 0 will cause signing to fail.
+ *  Out:     nonce32:   pointer to a 32-byte array to be filled by the function.
+ *  In:      msg32:     the 32-byte message hash being verified (will not be NULL)
+ *           key32:     pointer to a 32-byte secret key (will not be NULL)
+ *      xonly_pk32:     the 32-byte serialized xonly pubkey corresponding to key32
+ *                      (will not be NULL)
+ *           algo16:    pointer to a 16-byte array describing the signature
+ *                      algorithm (will not be NULL).
+ *           data:      Arbitrary data pointer that is passed through.
```

Comment:
Maybe we should make it more explicit that this additional entropy?
Or would it be better to have two separate arguments, one for additional entropy (maybe fixed length or with length description?) and one for arbitrary user data?

Reply:
That is up to the nonce function.

For example, there is a nonce function that uses the data function to exfiltrate the nonce itself for use in ECDH to generate stealth addresses.  For someone that needed very low latency signing a nonce function that took an argument to a datastructure of precomputed nonces.

Two arguments would address that-- e.g. one always for extra stuff, but OTOH, the singing function might end up with too many arguments that the user is always stuffing NULL into?   Maybe less of an issue if synthetic nonces are strongly encouraged.  But those could also be accomplished another way-- e.g. passing randomness into the context at init time.

Reply:
related: https://github.com/bitcoin-core/secp256k1/issues/757

> Two arguments would address that-- e.g. one always for extra stuff, but OTOH, the singing function might end up with too many arguments that the user is always stuffing NULL into? 

If there are too many NULLs, maybe we should just have a wrapper that does not allow specifying the nonce function. Specifying the nonce function is pretty advanced stuff anyway ("do this only if you know what you're doing").

Reply:
> Maybe we should make it more explicit that this additional entropy?
> Or would it be better to have two separate arguments [...] ?

I don't think so, as said above this is up to the nonce function. It's easy to put both randomness and your additional data in the data arg of your custom nonce function.
The doc for nonce data in schnorrsig_sign says `If it is non-NULL and secp256k1_nonce_function_bip340 is used, then ndata must be a pointer to 32-byte auxiliary randomness as per BIP-340.`

But I'd be fine with making schnorrsig_sign a function that doesn't take a nonce function argument and the noncedata arg is renamed to randomness (or something).
related: #589 which shows that the best way to implement sign-to-contract (and therefore anti-covert-channel mechanisms) is not via the nonce function. Therefore, changing the nonce function is probably rarely used compared to normal signing. Moreover, there's little benefit to keep this similar to the ecdsa_sign API because it already takes a keypair instead of a secret key byte array.

Reply:
> But I'd be fine with making schnorrsig_sign a function that doesn't take a nonce function argument and the noncedata arg is renamed to randomness (or something).

You mean remove the ability to pass custom functions entirely? Or have two sign functions, one with and without the feature? 

Yeah, maybe the user just doesn't need this at all. Anyway the nonce function is very fast, and if the user wants randomized signing and has some magic other entropy source for any reason, it's possible to pass entropy. Is there any other reason why the user would want this? 

Removing it would also leave us with more flexibility for keeping an additional counter in the signing context if we want to do something like `_sign_and_randomize` in the future (https://github.com/bitcoin-core/secp256k1/issues/780#issuecomment-670531851). (Otherwise, where would the counter go for user-defined nonce functions? "Nowhere" is a possible answer but maybe not a great one.).

Reply:
Two functions. The one without the feature gets to have the simplest `sign()` name.